### ISSUE TITLE
[Feature] sparse_tensor_dense_matmul with two SparseTensor

### ISSUE THREAD

**Issue by danielegrattarola** [tensorflow/tensorflow#10375]
Would it be possible to implement matrix multiplication between two `SparseTensor`? 

Currently, it's possible to multiply two dense `Tensor` with many zeros using the `X_is_sparse` parameter of `tf.matmul`, and a `SparseTensor` with a `Tensor` using `tf.sparse_tensor_dense_matmul`. 

However, with some datasets it is impossible to store dense matrices in memory and it would be great if we could have the option of fully sparse multiplication.

Thanks,
Daniele

**Comment by asimshankar** [comment]
@ebrevdo @rmlarsen : Do you know if anyone is planning to add this?

**Comment by wbwvos** [comment]
Another thing I would like to add is that it would be awesome if there would be an option to **return a `SparseTensor`** instead of a `Tensor`, I am doing normalization of a `SparseTensor` with `tf.sparse_tensor_dense_matmul` but this results in a dense `Tensor` with the same shape as the `SparseTensor`, which makes the whole `SparseTensor` obsolete.

**Comment by ebrevdo** [comment]
We are looking into sparse-sparse matmuls but won't have anything available
in the near future.  If we do, it won't be
called sparse_tensor_dense_matmul.

**Comment by YuanhaoGong** [comment]
Sparse matrix multiply another sparse matrix. The implementation should be simpler than "sparse X dense". It is so strange that the former does not exist in TF but later one exists.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by emkessler** [comment]
+1 on this request

**Comment by drabastomek** [comment]
+1 Here here!

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
@ebrevdo any updates on sparse-sparse matmul support?  Should this be "contributions welcome"?

**Comment by KonstZ** [comment]
+1

**Comment by plison** [comment]
+1

**Comment by ebrevdo** [comment]
Yes; at this point I'd say this is a "contributions welcome" since no one
is actively working on this.

**Comment by wuciawe** [comment]
+1

**Comment by hero78119** [comment]
+1

**Comment by everwind** [comment]
+1

**Comment by gongliyu** [comment]
+1

**Comment by jinhw1989** [comment]
+1

**Comment by pchoganwala** [comment]
+1

**Comment by jacarvalho** [comment]
+1

**Comment by danielegrattarola** [comment]
Hi, I'm just checking back after a few months. 
I am starting to look into developing this on my own, if anybody has any suggestions they are more than welcome (maybe @ebrevdo can help). I'm a total C++ noob but I'll try to edit some existing ops in order to reuse boilerplate code. 

Is [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/sparse_tensor_dense_matmul_op.cc) a good starting point (sparse_tensor_dense_matmul_op.cc/cu.cc/h)?
I was thinking that maybe all that's needed is to convert the sparse inputs to a format supported by cuSPARSE and let it handle the multiplication. Would that be a possible approach? 

Thanks

**Comment by ebrevdo** [comment]
We do have a sparse matrix TF library that uses cuSparse; and it does have
this functionality.  We hope to move it into tf.linalg.experimental
sometime this year.

**Comment by chaoyanghe** [comment]
class GraphConvolutionSparse(Layer):
	"""Graph convolution layer for sparse inputs."""

	def __init__(self, input_dim, output_dim, adj, features_nonzero, dropout=0., act=tf.nn.relu, **kwargs):
		super(GraphConvolutionSparse, self).__init__(**kwargs)
		with tf.variable_scope(self.name + '_vars'):
			self.vars['weights'] = weight_variable_glorot(input_dim, output_dim, name="weights")
		self.dropout = dropout
		self.adj = adj
		self.act = act
		self.issparse = True
		self.features_nonzero = features_nonzero

	def _call(self, inputs):
		x = inputs
		x = dropout_sparse(x, 1 - self.dropout, self.features_nonzero)
		x = tf.sparse_tensor_dense_matmul(x, self.vars['weights'])  # X W
		x = tf.sparse_tensor_dense_matmul(self.adj, x)  # A X W
		outputs = self.act(x)
		return outputs

I am working on GCN (Graph Convolutional Network). My graph data has more than 1 million nodes. When I run this model, I meet a memory explosion which will kill my process. But once I reduce the size of the graph, it works well. So my question is: is there any solution to the large-scale matrix multiplication? I am stuck here, please help me! Thanks!

**Comment by chaoyanghe** [comment]
> class GraphConvolutionSparse(Layer):
> """Graph convolution layer for sparse inputs."""
> 
> ```
> def __init__(self, input_dim, output_dim, adj, features_nonzero, dropout=0., act=tf.nn.relu, **kwargs):
> 	super(GraphConvolutionSparse, self).__init__(**kwargs)
> 	with tf.variable_scope(self.name + '_vars'):
> 		self.vars['weights'] = weight_variable_glorot(input_dim, output_dim, name="weights")
> 	self.dropout = dropout
> 	self.adj = adj
> 	self.act = act
> 	self.issparse = True
> 	self.features_nonzero = features_nonzero
> 
> def _call(self, inputs):
> 	x = inputs
> 	x = dropout_sparse(x, 1 - self.dropout, self.features_nonzero)
> 	x = tf.sparse_tensor_dense_matmul(x, self.vars['weights'])  # X W
> 	x = tf.sparse_tensor_dense_matmul(self.adj, x)  # A X W
> 	outputs = self.act(x)
> 	return outputs
> ```
> 
> I am working on GCN (Graph Convolutional Network). My graph data has more than 1 million nodes. When I run this model, I meet a memory explosion which will kill my process. But once I reduce the size of the graph, it works well. So my question is: is there any solution to the large-scale matrix multiplication? I am stuck here, please help me! Thanks!

PS: the X inputs in the code is super large.

**Comment by jinhw1989** [comment]
@chaoyanghe Hmm, facing the exact same problem on GCN.

**Comment by GeorgieJi** [comment]
+1. I'm facing same problem on GCN . Although the memory of my GPU is 24gb, it's still not enough to apply a dense tensor of adjacent matrix to execute calculation.

**Comment by chpohl** [comment]
@danielegrattarola Is there any progress on this? I also need this functionality and would be willing to assist you in developing it. I also have some prior experience with developing new tf-ops. Let me know if I can assist you

**Comment by danielegrattarola** [comment]
@chpohl I gave up after @ebrevdo told me that they already have the functionality implemented, but it never got released.
Unfortunately, TF2 and the new Keras broke most practical uses of sparse matrices, so I don't think the the tf devs have any interest in supporting the feature moving forward.

If you want to implement sparse-sparse matmul for some older version of TF drop me an email, maybe we can look at it together.

**Comment by ebrevdo** [comment]
It's available in tf2 nightlies under the tf.linalg.experimental module

**Comment by danielegrattarola** [comment]
Thanks @ebrevdo! I'll close the issue.

**Comment by chpohl** [comment]
I just install tf-nightly-gpu and cannot find a tf.linalg.experimental module. Do you mean the functionality under tensorflow.python.ops.linalg.sparse ? When I read it correctly, the matmul there supports both sparse inputs.
